library/ecdsa.c

```diff
@@ -255,7 +255,9 @@ static int ecdsa_sign_restartable( mbedtls_ecp_group *grp,
     mbedtls_mpi *pk = &k, *pr = r;
 
     /* Fail cleanly on curves such as Curve25519 that can't be used for ECDSA */
-    if( grp->N.p == NULL )
+    if( grp->id == MBEDTLS_ECP_DP_CURVE25519 ||
```

Comment:
Could we make the `mbedtls_ecp_group_id` enumeration explicit in a way that Weierstrass curves can be distinguished from non-Weierstrass ones by checking a single bit?

Reply:
@hanno-arm It is a good idea, but I don't see how could we do it without breaking the ABI.

@gilles-peskine-arm I think it would be handy and improve readability, but do you think it is in scope for this PR?

Reply:
I think rather we (Arm) should do it, and then this PR should use it. But I don't make it a hard requirement: it would be better if to do it this way, but if it takes too long we can leave it for later.

Reply:
The curves in question are currently numbered 9 and 13 (dec). We could check `grp->id & 0xFFFFFFFFFB == 0xFFFFFFFFFB`, but I don't think that would be much faster than the `switch` in `mbedtls_ecdsa_can_do`.

Reply:
I think this is pretty much the last outstanding issue for this PR. Should we at least move the existing enumeration into its own function so that it's at least consistent across uses?

Reply:
@gilles-peskine-arm I think it would take too much time to do it like that. Are you happy with raising this as an issue and resolving it some day after this PR has been merged?

Reply:
Yes, that's fine by me.

Reply:
Tracking it here: https://github.com/ARMmbed/mbedtls/issues/2375

Reply:
Great, many thanks!